Let Insane difficulty start with gates

aco4 opened the pull request.

aco4 committed:
Let Insane difficulty start with gates

KJeff01 commented:
I'd bet if you passed just the structure's stattype to that function there would be a huge perf increase.

During some tests many years ago, I noticed passing whole _game objects_ as arguments incurs a somewhat "big" performance cost and even more for arrays of them. The only thing (much) worse is using enumDroid|Struct|Feature a lot... which isn't always avoidable.

After that I went out of my way to somewhat overuse getObject() in many parts of Cobra to shave off several milliseconds of time from the worst perf times in the game logs.

Not that it matters much most of the time since it's not _that_ serious passing game objects, and this case happens only once. Still, potentially hundreds or thousands of structures are going to be scooped up at once in this at the start of every skirmish/mp session. And the side effect of this may be rather big on weaker computers.

aubergineagain commented:
Suggestion:

```js
function removeStructs(player)
{
    const difficulty = playerData[player].difficulty;

    for (const structure of enumStruct(player))
	{
        if (!keepStructure(structure.stattype, difficulty))
		{
			removeObject(structure);
		}
	}
}

// CAMP_CLEAN on INSANE
const keepCleanOnInsane = {
    WALL: true,
    DEFENSE: true,
    GATE: true,
    RESOURCE_EXTRACTOR: true,
}

// CAMP_BASE on INSANE
const keepBaseOnInsane = {
    CYBORG_FACTORY: true,
    COMMAND_CONTROL: true,
}

// CAMP_BASE non-INSANE
const keepBaseNotInsane = {
    WALL: true,
    DEFENSE: true,
    GATE: true,
    CYBORG_FACTORY: true,
    COMMAND_CONTROL: true,
}

/**
 * @param {number} stattype
 * @param {number} difficulty
 */
function keepStructure(stattype, difficulty)
{
	if (baseType === CAMP_CLEAN && difficulty !== INSANE)
	{
		return false; // remove everything
	}
	if (baseType === CAMP_CLEAN && difficulty === INSANE)
	{
		return keepCleanOnInsane[stattype] || false;
	}
	else if (baseType === CAMP_BASE && difficulty !== INSANE)
	{
		return keepBaseNotInsane[stattype] || false;
	}
	else if (baseType === CAMP_BASE && difficulty === INSANE)
	{
		return keepBaseOnInsane[stattype] || false;
	}

    return true; // keep everything
}
```

aco4 committed:
Optimize performance

Pass stattype instead of entire structure object, which may improve performance as per @KJeff01. Also increase the verbosity of the function names to help mitigate global scope pollution

KJeff01 commented:
Seems good to me.

aco4 commented:
> During some tests many years ago, I noticed passing whole game objects as arguments incurs a somewhat "big" performance cost and even more for arrays of them. The only thing (much) worse is using enumDroid|Struct|Feature a lot... which isn't always avoidable.

@KJeff01 I think this only applies when passing objects to/from the WZ API backend, due to marshalling from JS <--> C++. It should completely fine to pass large objects within the script-side of things. Do correct me if I'm wrong.
